Add `textEnum` column builder and `EnumNameConverter`

ValentinVignal opened the pull request:
Should close https://github.com/simolus3/drift/issues/2115

ValentinVignal committed:
:sparkles: textEnum

ValentinVignal committed:
:white_check_mark: Update the tests

ValentinVignal commented on drift_dev/lib/src/analysis/resolver/drift/table.dart:
```diff
@@ -72,6 +73,9 @@ class DriftTableResolver extends LocalElementResolver<DiscoveredDriftTable> {
             (msg) => reportError(
                 DriftAnalysisError.inDriftFile(column.definition ?? stmt, msg)),
             dartClass.classElement.thisType,
+            column.type.type == BasicType.int
+                ? EnumType.intEnum
+                : EnumType.textEnum,
```
Is it the correct way to do this conditional check?

simolus3 replied:
Let's use the local `type` variable instead. You can compare it against `DriftSqlType.int`.

ValentinVignal replied:
I've used `type` in [🎨 Better type comparison and matcher](https://github.com/simolus3/drift/pull/2168/commits/4030db4c5f6eb7795f3f21c2a65065f025bdbdc0)

ValentinVignal commented on drift_dev/test/analysis/resolver/dart/enum_columns_test.dart:
```diff
@@ -55,8 +64,14 @@ void main() {
       file.allErrors,
-      contains(isA<DriftAnalysisError>().having((e) => e.message, 'message',
-          allOf(contains('Not an enum'), contains('NotAnEnum')))),
+      containsAllInOrder(
+        [
+          isA<DriftAnalysisError>().having((e) => e.message, 'message',
+              allOf(contains('Not an enum'), contains('NotAnEnum'))),
+          isA<DriftAnalysisError>().having((e) => e.message, 'message',
+              allOf(contains('Not an enum'), contains('NotAnEnum'))),
+        ],
+      ),
     );
```
Do you know a better way to verify an iterable contains 2 occurrences of a matcher?

simolus3 replied:
I would just extract the matcher into a local variable and use it twice:

```dart
final notAnEnumError = isDriftError(allOf(contains('Not an enum'), contains('NotAnEnum')));

expect(file.allErrors, [notAnEnumError, notAnEnumError);
```

ValentinVignal replied:
I extracted the matcher in [🎨 Better type comparison and matcher](https://github.com/simolus3/drift/pull/2168/commits/4030db4c5f6eb7795f3f21c2a65065f025bdbdc0)

simolus3 reviewed (commented):
The implementation looks good to me, nice work! I have some minor comments on how to compare the type and for the matcher.

We should also mention this in the changelog and the documentation, but I can take care of that as well.

ValentinVignal committed:
:art: Better type comparison and matcher

ValentinVignal committed:
:arrow_up: Upgrade version number

ValentinVignal committed:
:memo: Update the documentation

ValentinVignal committed:
Merge remote-tracking branch 'upstream/develop' into textEnum

# Conflicts:
#	drift/CHANGELOG.md
#	drift_dev/CHANGELOG.md

simolus3 commented on drift_dev/pubspec.yaml:
```diff
@@ -1,6 +1,6 @@
 name: drift_dev
 description: Dev-dependency for users of drift. Contains a the generator and development tools.
-version: 2.3.0
+version: 2.4.0-dev
```
Please also change the `drift: ` dependency in this package to `>=2.4.0 <2.5.0`.

simolus3 reviewed (commented):
Thanks for also updating the documentation!

In the generator, it would also be good to have a test for drift files - you can take a look at the `parses enum columns` test in `drift_dev/test/analysis/resolver/drift/table_test.dart` file for inspiration.
There should also be tests interacting with `EnumNameConverter` at runtime - similar to `drift/test/database/types/enum_index_converter.dart`.

ValentinVignal committed:
Merge remote-tracking branch 'upstream/develop' into textEnum

# Conflicts:
#	drift_dev/CHANGELOG.md
#	drift_dev/pubspec.yaml

ValentinVignal committed:
:bug: Use correct type for enum converter

ValentinVignal committed:
:label: Add todo status

ValentinVignal committed:
:white_check_mark: Update drift tests

ValentinVignal committed:
:white_check_mark: Verify the EnumNameConverter is used

ValentinVignal committed:
Merge remote-tracking branch 'upstream/develop' into textEnum

ValentinVignal committed:
:bulb: Better test description

ValentinVignal committed:
:white_check_mark: Add enum name converter test

ValentinVignal committed:
:sparkles Add ENUMNAME type

ValentinVignal commented:
> In the generator, it would also be good to have a test for drift files - you can take a look at the `parses enum columns` test in `drift_dev/test/analysis/resolver/drift/table_test.dart` file for inspiration.

I've introduced `ENUMNAME` in [:sparkles Add ENUMNAME type](https://github.com/simolus3/drift/pull/2168/commits/3cfa69c9d965cc2667603580590ca020a6424c90). But I don't use this feature in my apps so I don't really know what I'm doing here... Please be careful when you review this part

> There should also be tests interacting with `EnumNameConverter` at runtime - similar to `drift/test/database/types/enum_index_converter.dart`.

I've added those tests in [✅ Add enum name converter test](https://github.com/simolus3/drift/pull/2168/commits/ed82e46f2dfffa9998031df038f2d8de384d84a7)

simolus3 committed:
Add docs for drift files

simolus3 reviewed (approved):
Thanks a lot for your contribution, I think this is a helpful feature for sure!
